proposed/dotnet-sdk-unified-config.md

```diff
@@ -0,0 +1,150 @@
+# A unified configuration file and experience for the .NET SDK
+
+**Owner** [Chet Husk](https://github.com/baronfel) | [Daniel Plaisted](https://github.com/dsplaisted)
+
+The .NET SDK is a product that consists of a number of different tools, all owned by different teams, all operating at slightly different levels of interaction. Some tools operate on the entire repository, some tools on a single project, and yet other tools on a single file.
+
+Because of the disparate nature of these tools, the configuration for each tool is different. This makes it difficult for users to understand how to configure the tools, and difficult for the tools to share configuration information. This document proposes a single, centralized document for repository-wide configuration that tools belonging to the SDK (as well as SDK-adjacent 1st- and 3rd-party tools) can use to share configuration information in a way that has
+
+* a consistent user experience
+* a well supported IDE experience
+* is shareable across team members and trackable in source control
+
+## What exists today
+
+Today there are several places where teams have put configuration for repo- or project-level configuration for tooling:
+
+* global.json files
+* NuGet.config
+* .config/dotnet-tools.json
+* .runsettings (dotnet test)
+
+In addition, several tools are only configurable via environment variables, which are not easily shareable or trackable in source control (due to the lack) of support for .env or other such conventions in the `dotnet` CLI.
+
+All of these files and systems arose independently, have different formats, and the amount of tooling support for them varies significantly. We have 'shipped the org structure' in terms of our configuration.
+
+## What is the need?
+
+There are [several dotnet/sdk items](https://github.com/dotnet/sdk/issues?q=label%3Aconfig-system-candidate) where the concept of a unified configuration system would be beneficial. These include:
+
+* expressing user behavioral preferences around build/pack/publish Configurations
+* if binlogs should be made when performing builds
+* the use or non-use of advanced terminal behaviors for build
+* default verbosities for build/pack/publish
+* custom aliases for CLI commands
+* unified opt-out of telemetry
+* storing tooling decisions like which mode `dotnet test` should execute in
+
+All kinds of repository-wide configuration need a unified place to be expressed that we lack today.
```

Comment:
Would it be better to extend global.json instead of introducing yet another file?

Reply:
Worth having the conversation, but global.json first and foremost belongs to the runtime today. I'd like to have something that tooling can work with that has guaranteed zero impact on runtime scenarios.

There is a bit of https://xkcd.com/927/ in this proposal, though.

Reply:
> global.json first and foremost belongs to the runtime today

From my point of view, global.json belongs to the SDK. There is a tiny bit of code in in dotnet/runtime that reads it to find the version of SDK to call. This code only ever runs when `dotnet` is used to invoke SDK.

Also, different parts of SDK or SDK extensions have started to use global.json to store random configuration. Check what's in global.json in the dotnet/sdk repo: https://github.com/dotnet/sdk/blob/main/global.json